gobblin-tunnel/src/main/java/gobblin/tunnel/Tunnel.java

```diff
@@ -0,0 +1,178 @@
+ *    externally hosted resource, it should check if the WorkUnit has a proxy host and port defined.
+ * 2. If a proxy is defined, it should extract the remote host and port from the target URL hosting the resource (e.g.
+ *    the JdbcProvider gets this from the connectionUrl.)
+ * 3. The extractor then creates a Tunnel instance configured with the remote host and port and the proxy host and port.
+ * 4. The Tunnel starts a thread that listens on an arbitrary port on localhost.
+ * 5. The extractor then points the target URL to the localhost and port the Tunnel is listening on. (E.g. in the case
+ *    of JDBC, the JdbcProvider changes the connectionUrl to replace the remote host and port with the localhost and
+ *    port before passing it on to the driver.)
+ * 6. Hence when the extractor client (e.g. JDBC driver) creates a connection, it connects to the Tunnel socket instead
+ *    of the actual target host.
+ * 7. The Tunnel then connects to the remote host through the proxy via a HTTP CONNECT request.
+ * 8. If established successfully, the Tunnel then simply relays bytes back and forth between the Gobblin extractor and
+ *    the target host via the intermediate proxy.)
+ * 7. When the Gobblin extractor (e.g. JDBC data source) is closed down, the Tunnel must be shut down as well.
+ *
+ * The Tunnel can accept as many connections as the JdbcExtractor opens. It uses NIO to minimize resource usage.
+ *
+ * @author navteniev@linkedin.com
+ * @author kkandekar@linkedin.com
+ */
+public class Tunnel {
+  public static final int NON_EXISTENT_PORT = -1;
+
+  private static final Logger LOG = LoggerFactory.getLogger(Tunnel.class);
+
+  private ServerSocketChannel server;
+  private Thread thread;
+  private final Config config;
+
+  private Tunnel(String remoteHost, int remotePort, String proxyHost, int proxyPort) {
+    config = new Config(remoteHost, remotePort, proxyHost, proxyPort);
+  }
+
+  private Tunnel open() throws IOException {
+    try {
+      server = ServerSocketChannel.open().bind(null);
+      startTunnelThread();
+
+      return this;
+    } catch (IOException ioe) {
+      LOG.error("Failed to open the tunnel", ioe);
+      throw ioe;
+    }
+  }
+
+  public int getPort() throws IOException {
+    SocketAddress localAddress = null;
+
+    try {
+      if (server != null && server.isOpen()) {
+        localAddress = server.getLocalAddress();
+      }
+      if (localAddress instanceof InetSocketAddress) {
+        return ((InetSocketAddress) localAddress).getPort();
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to get tunnel port", e);
+      throw e;
+    }
+    return NON_EXISTENT_PORT;
+  }
+
+  private void startTunnelThread() {
+    thread = new Thread(new Dispatcher(), "Tunnel Listener");
+    thread.start();
+  }
+
+  public boolean isTunnelThreadAlive() {
+    return (thread != null && thread.isAlive());
+  }
+
+  private class Dispatcher implements Runnable {
```

Comment:
True, it's better for exceptions to be propagated to the caller, but in this case, the exception may occur in the middle of a long-ish running thread (e.g. remote server or proxy closes socket), which means it is actually occurring somewhere other the caller. That is, the exception may occur long after JdbcProvider calls Tunnel.open() or long before it calls Tunnel.close()  -- e.g. when fetching the query results. Hence it's not possible for the Tunnel to log the exception at an accurate place within the context of the rest of the Extractor.

However, when this happens, the JdbcExtractor would also inevitably experience an exception as the JDBC driver fails to communicate with the Tunnel, so that would be a more accurate place for the exception to be logged.

As for resource clean-up, when a thread throws an un-checked exception it simply exits. It won't be logged, which is a shortcoming, but it won't cause a hung thread. It's possible to handle that using  UncaughtExceptionHandler, but at that point I guess it may be better to use ExecutorService. I'll try that out.